python/python/lance/dataset.py

```diff
@@ -191,6 +191,12 @@ def __setstate__(self, state):
         self._uri, version = state
         self._ds = _Dataset(self._uri, version)
 
+    def __copy__(self):
+        ds = LanceDataset.__new__(LanceDataset)
+        ds._uri = self._uri
+        ds._ds = copy.copy(self._ds)
+        return ds
```

Comment:
We call `copy.copy()` in `checkout_version`. We needed this change to make sure that we don't call `__init__()` on the `LanceDataset` and reinstantiate without all the storage options kept.